plugins/default/src/Ide/Plugin/ModuleName.hs

```diff
@@ -0,0 +1,162 @@
+                                             GhcMonad (getSession),
+                                             HsModule (hsmodName),
+                                             ParsedModule (pm_parsed_source),
+                                             SrcSpan (RealSrcSpan), unLoc)
+import           GhcPlugins                 (HscEnv (hsc_IC),
+                                             InteractiveContext (ic_dflags))
+import           Ide.Types                  (CommandFunction, CommandId (..),
+                                             PluginCommand (..),
+                                             PluginDescriptor (..),
+                                             PluginId (..),
+                                             defaultPluginDescriptor)
+import           Language.Haskell.LSP.Core  (LspFuncs, getVirtualFileFunc)
+import           Language.Haskell.LSP.Types (ApplyWorkspaceEditParams (..),
+                                             CAResult (CACodeAction),
+                                             CodeAction (CodeAction),
+                                             CodeActionKind (CodeActionQuickFix),
+                                             CodeLens (CodeLens),
+                                             CodeLensParams (CodeLensParams),
+                                             Command (Command),
+                                             ServerMethod (..),
+                                             TextDocumentIdentifier (TextDocumentIdentifier),
+                                             TextEdit (TextEdit), Uri,
+                                             WorkspaceEdit (..),
+                                             uriToNormalizedFilePath)
+import           Language.Haskell.LSP.VFS   (virtualFileText)
+import           System.FilePath            (dropExtension)
+
+-- |Plugin descriptor
+descriptor :: PluginId -> PluginDescriptor
+descriptor plId =
+  (defaultPluginDescriptor plId)
+    { pluginId = plId,
+        pluginCodeLensProvider = Just codeLens
+       ,pluginCommands = [PluginCommand editCommandName editCommandName editCmd]
+        -- pluginCodeActionProvider = Just codeAction
+    }
+
+-- | Generate code lenses
+codeLens :: LspFuncs c -> IdeState -> PluginId -> CodeLensParams -> IO (Either a2 (List CodeLens))
+codeLens lsp state pluginId (CodeLensParams (TextDocumentIdentifier uri) _) = do
+  pid <- getPid
+  actions (asCodeLens (mkLspCmdId pid pluginId editCommandName)) lsp state uri
+
+-- | Generate code actions.
+-- NOTE: Not invoked on an empty module (but codeLens is, why?)
+codeAction :: LspFuncs c -> IdeState -> p1 -> TextDocumentIdentifier -> p2 -> p3 -> IO (Either a (List CAResult))
+codeAction lsp state _plId (TextDocumentIdentifier uri) _range _ = actions asCodeAction lsp state uri
+
+-- Copied from "Ide.Plugin"
+mkLspCmdId :: T.Text -> PluginId -> CommandId -> T.Text
+mkLspCmdId pid (PluginId plid) (CommandId cid)
+  = pid <> ":" <> plid <> ":" <> cid
```

Comment:
I'm not sure about this copy-paste. Is there any way to avoid it?

Reply:
yes, by exposing the function in Ide.Plugin

Reply:
Any reason not to?

Reply:
no idea, should I do it?

Reply:
sure! it could be useful for others plugins

Reply:
done